sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java

```diff
@@ -359,6 +368,27 @@ as we can guarantee we only need at most two buffers for any call to write (two
         }
     }
 
+    private Mono<Response<BlockBlobItem>> determineUploadFullOrChunked(final Flux<ByteBuffer> data,
+        final Function<Flux<ByteBuffer>, Mono<Response<BlockBlobItem>>> path1,
+        final BiFunction<Flux<ByteBuffer>, Long, Mono<Response<BlockBlobItem>>> path2) {
+        final int[] predicateCount = { 4 * Constants.MB };
+        final boolean[] usePath1 = { true };
+
+        return data
+            .filter(ByteBuffer::hasRemaining)
+            .doOnEach(signal -> usePath1[0] = !signal.isOnComplete())
+            .bufferUntil(buffer -> {
+                predicateCount[0] -= buffer.remaining();
+                return predicateCount[0] < 0;
+            }, true)
+            .next()
+            .flatMap(bufferedData -> usePath1[0]
+                ? path1.apply(data)
```

Comment:
I'm not an expert with Reactor, but I don't know if it's valid to pass `data` to `path1` after already reading some of the buffers.  This will work with some Fluxes (like `Flux.just()`), but it may not work with an arbitrary Flux which I believe is free to discard buffers after they have been read.

For example, I believe the default mode for a `Flux` received from `reactor-netty` will recycle memory as soon as a buffer has been published.  You would need to call `retain()` to change this behavior.

https://projectreactor.io/docs/netty/release/api/reactor/netty/ByteBufFlux.html#retain--

Reply:
In our SDK, the contract in general when user provide `Flux<ByteBuffer>` is - the provided Flux has to be replayable means multiple subscriptions can be done.

Reactor Netty's design is based on pooled ByteBuf (ref counted) so as soon as each chunk is consumed and if ref count is zero, the chunk is returned to the pool.

Reply:
This seems problematic to me for a few reasons:

1. To upload a large file (say 1 GB) using `upload(Flux<ByteBuffer>)`, the entire content needs to be stored in memory until the upload is complete.
2. If the provided `Flux<ByteBuffer>` is not replayable, the failure mode could be really bad (e.g. uploading the wrong content).

Reply:
Just to clarify a couple points here: Anything passed to generated code has to be replayable. The method under discussion, though, is for the express purpose of uploading non-replayable Fluxes (also for Fluxes of unknown length, but not as important here).